scipy/integrate/tests/test_quadrature.py

```diff
@@ -457,3 +459,398 @@ def func(x):
+        assert res == ref
+
+        # Now reduce `maxfun`. We won't meet tolerances.
+        f.feval, f.calls = 0, 0
+        maxfun -= 1
+        res = _tanhsinh(f, 0, f.b, maxfun=maxfun)
+        assert self.error(res.integral, f.ref) < res.error > default_atol
+        assert res.feval == f.feval < ref.feval
+        assert f.calls == ref.calls - 1
+        assert res.success is False
+        assert res.status == 2
+        assert res.message.endswith("evaluation limit to be exceeded.")
+
+        # Take this result to be the new reference
+        ref = res
+        ref.calls = f.calls
+
+        # Test `atol`
+        f.feval, f.calls = 0, 0
+        # With this tolerance, we should get the exact same result as ref
+        atol = np.nextafter(ref.error, np.inf)
+        res = _tanhsinh(f, 0, f.b, atol=atol)
+        assert res.integral == ref.integral
+        assert res.error == ref.error
+        assert res.feval == f.feval == ref.feval
+        assert f.calls == ref.calls
+        # Except the result is considered to be successful
+        assert res.success is True
+        assert res.status == 0
+        assert res.message.startswith("The algorithm completed successfully")
+
+        f.feval, f.calls = 0, 0
+        # With a tighter tolerance, we should get a more accurate result
+        atol = np.nextafter(ref.error, -np.inf)
+        res = _tanhsinh(f, 0, f.b, atol=atol)
+        assert self.error(res.integral, f.ref) < res.error < atol
+        assert res.feval == f.feval > ref.feval
+        assert f.calls > ref.calls
+        assert res.success is True
+        assert res.status == 0
+        assert res.message.startswith("The algorithm completed successfully")
+
+        # Test `rtol`
+        f.feval, f.calls = 0, 0
+        # With this tolerance, we should get the exact same result as ref
+        rtol = np.nextafter(ref.error/ref.integral, np.inf)
+        res = _tanhsinh(f, 0, f.b, rtol=rtol)
+        assert res.integral == ref.integral
+        assert res.error == ref.error
+        assert res.feval == f.feval == ref.feval
+        assert f.calls == ref.calls
+        # Except the result is considered to be successful
+        assert res.success is True
+        assert res.status == 0
+        assert res.message.startswith("The algorithm completed successfully")
+
+        f.feval, f.calls = 0, 0
+        # With a tighter tolerance, we should get a more accurate result
+        rtol = np.nextafter(ref.error/ref.integral, -np.inf)
+        res = _tanhsinh(f, 0, f.b, rtol=rtol)
+        assert self.error(res.integral, f.ref)/f.ref < res.error/res.integral < rtol
+        assert res.feval == f.feval > ref.feval
+        assert f.calls > ref.calls
+        assert res.success is True
+        assert res.status == 0
+        assert res.message.startswith("The algorithm completed successfully")
+
+    @pytest.mark.parametrize('rtol', [1e-4, 1e-14])
+    def test_log(self, rtol):
```

Comment:
Tighten tolerances in this test. The only difference should be due to loss of precision due to log scale.